questions/000-fn-pointer-vs-fn-type.md

```diff
@@ -0,0 +1,36 @@
+Answer: 4400
+Difficulty: 2
+
+# Hint
+
+The integer type parameters for `a`  have no affect in this example - they are a
+red herring.
+
+# Explanation
+
+There is no difference in size between any of `a::<u8>`,  `a::<u16>`, `a::<u32>`
+and `a::<u64>`. In fact all `fn` items have zero-sized types (ZST). To see why
+this makes sense, remember that the size of a closure is the combined size of
+its captured variables. `fn` items are not closures and cannot capture variables
+but they are interchangeable with non-capturing closures so it makes sense that
```

Comment:
I agree that it makes sense but I think this is only half the story; there are disadvantages to this representation, and other representations may also make sense for other reasons that outweigh this one. It would be good to give some intuition for why Rust works this way with fn items. AIUI this treatment is necessary in order for things like `iter.filter(f).map(m)` to monomorphize in a zero cost way.

Reply:
The way I see it, if the size is > 0, what _exactly_ would be copied/moved around in memory when you pass the function around? It's obviously not the executable code of the function. It could be a pointer to the executable code, but that isn't necessary because each function has a unique type so the compiler can just insert the jump where the function actually ends up being called. This seems to fit well with what you said about zero cost monomorphisation in iterator combinators.

Reply:
The whole point of this quiz question would be to convey that function values have unique types, unlike in every other language. I consider the primary responsibility of the explanation to be justifying why function values have unique types unlike every other language in order for the reader to really learn what's going on. I think "because each function has a unique type" doesn't get to this.

Reply:
That was sort of the point I was getting to. A unique type per function allows the compiler to identify a function at compile-time and simply write the instructions that call it at the sites where it is eventually called.

Reply:
Yep, sounds good. My holdup was that between the two inverse perspectives,

- function values are zero sized (free of runtime representation) because they have unique types; *vs*
- function values have unique types because the compile-time optimizations that are required for meeting Rust's zero cost abstraction target require them to be free of runtime representation

the second is the more enlightening and accurate framing.